clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +243,461 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+        }
+
+        commitResult.whenComplete((result, error) -> {
+
+            if (error != null) {
+                // Commit request failed (commit request manager internally retries on
+                // retriable errors, so at this point we assume this is non-retriable, but
+                // proceed with the revocation anyway).
+                log.debug("Commit request before revocation failed with non-retriable error. Will" +
+                        " proceed with the revocation anyway.", error);
+            }
+
+            CompletableFuture<Void> userCallbackResult = invokeOnPartitionsRevokedCallback(revokedPartitions);
+            userCallbackResult.whenComplete((callbackResult, callbackError) -> {
+                if (callbackError != null) {
+                    log.error("User provided callback failed on invocation of onPartitionsRevoked" +
+                            " for partitions {}", revokedPartitions, callbackError);
+                    revocationResult.completeExceptionally(callbackError);
+                } else {
+                    revocationResult.complete(null);
+                }
+
+            });
+        });
+        return revocationResult;
+    }
+
+    /**
+     * Mark partitions as 'pending revocation', to effectively stop fetching while waiting for
+     * the commit offsets request to complete, and ensure the application's position don't get
+     * ahead of the committed positions. This mark will ensure that:
+     * <ul>
+     *     <li>No new fetches will be sent out for the partitions being revoked</li>
+     *     <li>Previous in-flight fetch requests that may complete while the partitions are being revoked won't be processed.</li>
+     * </ul>
+     */
+    private void markPendingRevocationToPauseFetching(Set<TopicPartition> partitionsToRevoke) {
+        // When asynchronously committing offsets prior to the revocation of a set of partitions, there will be a
+        // window of time between when the offset commit is sent and when it returns and revocation completes. It is
+        // possible for pending fetches for these partitions to return during this time, which means the application's
+        // position may get ahead of the committed position prior to revocation. This can cause duplicate consumption.
+        // To prevent this, we mark the partitions as "pending revocation," which stops the Fetcher from sending new
+        // fetches or returning data from previous fetches to the user.
+        log.debug("Marking partitions pending for revocation: {}", partitionsToRevoke);
+        subscriptions.markPendingRevocation(partitionsToRevoke);
+    }
+
+    private CompletableFuture<Void> invokeOnPartitionsRevokedCallback(Set<TopicPartition> partitionsRevoked) {
+        Optional<ConsumerRebalanceListener> listener = subscriptions.rebalanceListener();
+        if (listener.isPresent()) {
+            throw new UnsupportedOperationException("User-defined callbacks not supported yet");
```

Comment:
Yes, we'll have to resolve how the callbacks fit into this model that uses `Future`s, because the callbacks need to be invoked on the application thread.

Reply:
Interesting :)